hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -19,29 +38,232 @@ impl TraceDecoder for YkPTTraceDecoder {
+                    // We can only know the successor via a TIP update in a packet. Fall through to
+                    // `self.seek_tip()`.
+                    self.cur_loc = ObjLoc::OtherObjOrUnknown;
+                }
+            }
+        }
+
+        // If we get here then we can't statically figure out where to go next. A packet is going
+        // to have to tell us what to do.
+        self.seek_tip()
+    }
+
+    /// Keep decoding packets until we have some TNT decisions buffered.
+    fn seek_tnt(&mut self) -> Result<bool, HWTracerError> {
+        while self.tnts.is_empty() {
+            let pkt = self.packet()?; // Potentially populates `self.tnts`.
+            if pkt.is_none() {
+                return Ok(false); // Packet stream exhausted.
+            }
+        }
+        Ok(true)
+    }
+
+    /// Keep decoding packets until we receive a TIP update.
+    fn seek_tip(&mut self) -> Result<Option<Block>, HWTracerError> {
+        debug_assert_eq!(self.cur_loc, ObjLoc::OtherObjOrUnknown);
+        loop {
+            if let Some(_) = self.packet()? {
+                if let ObjLoc::MainObj(off) = self.cur_loc {
+                    return Ok(Some(self.lookup_block_from_main_bin_offset(off)));
+                }
+            } else {
+                return Ok(None); // No more packets.
+            }
+        }
+    }
+
+    /// Fetch the next packet and update iterator state.
+    fn packet(&mut self) -> Result<Option<Packet>, HWTracerError> {
+        let ret = if let Some(pkt_or_err) = self.parser.next() {
+            let pkt = pkt_or_err?;
+            if pkt.kind() == PacketKind::TIPPGE {
+                debug_assert!(!self.pge);
+                self.pge = true;
+            }
+            if pkt.kind() == PacketKind::TIPPGD {
+                debug_assert!(self.pge);
+                self.pge = false;
+            }
+            if let Some(vaddr) = pkt.target_ip() {
+                if self.pge {
+                    match code_vaddr_to_off(vaddr) {
+                        Some((obj, off)) if obj == *SELF_BIN_PATH => {
+                            self.cur_loc = ObjLoc::MainObj(off)
+                        }
+                        _ => self.cur_loc = ObjLoc::OtherObjOrUnknown,
+                    }
+                    self.tnts.clear();
+                }
+            }
+            if let Some(bits) = pkt.tnts() {
+                self.tnts.extend(bits);
+            }
+            Ok(Some(pkt))
+        } else {
+            Ok(None)
+        };
+        ret
+    }
 }
 
 impl<'t> Iterator for YkPTBlockIterator<'t> {
     type Item = Result<Block, HWTracerError>;
 
     fn next(&mut self) -> Option<Self::Item> {
         if !self.errored {
-            // FIXME: For now this is dummy code to prevent dead-code warnings. Later block binding
-            // logic will go here.
-            self.parser.next().unwrap().unwrap();
+            match self.do_next() {
+                Ok(None) => None,
```

Comment:
In general I suggest order `match` arms by the expected frequency, as the optimiser is somewhat likely to respect that. So `Ok(Some(blk))` should probably come first.

Reply:
Fixed in dd3bdd3dbec19c97e64116206b00b5c8ac4789a5